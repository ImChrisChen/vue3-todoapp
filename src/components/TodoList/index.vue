<template>
  <div class="todo-list">
    <input
      class="input"
      data-casemark="pre-add-input"
      placeholder="请输入要完成的事项"
      style="background: seashell; color: #2c3e50; display: inline-block; border-radius: 4px; padding: 10px"
      v-model="formData.name"
      @keydown.enter="handleAddTodo"
      type="text"
    />
    <button class="button" data-casemark="submit" @click="handleAddTodo">添加</button>

    <ul data-casemark="todolist" v-if="todoList.length > 0">
      <li
        v-for="(todo, i) in todoList"
        :class="{ 'todo-done': todo.done, todo: true }"
        :key="i"
        :data-casemark-id="todo.id"
      >
        <input class="input" type="text" v-model="todo.name" @change="handleUpdateTodo(todo)" />
        <span class="done" @click="handleUpdateDoneTodo(todo)">✔️</span>
        <span class="delete" @click="handleDeleteTodo(todo)">❌</span>
      </li>
    </ul>
    <div v-else style="margin-top: 30px">添加一个TODO吧</div>

    <button class="button" @click="handleVisibleDone">{{ isVisibleDone ? '隐藏' : '显示' }} 已完成</button>
  </div>
</template>

<script lang="ts" setup name="TodoList">
import { onMounted, reactive, ref, watch, watchEffect } from 'vue'
import { ElMessage } from 'element-plus'

import { Todo, todoService } from '@/services/todo.service'

const todoList = ref<Todo[]>([])
const originTodoList = ref<Todo[]>([])
const formData = reactive({ name: '' })
const isVisibleDone = ref(true)

onMounted(() => {
  todoService.getTodoList().then((list) => {
    originTodoList.value = [...list]
  })
})

// 完成的todo自动移入到最后
watchEffect(() => {
  // TODO 自定义排序规则
  let tempArr: Todo[] = []
  let ids = originTodoList.value
    .map((item) => {
      if (item.done) {
        return item.id
      }
      return null
    })
    .filter((v) => v)

  ids.forEach((id) => {
    let index = originTodoList.value.findIndex((item) => item.id && item.id === id)
    let item = originTodoList.value.splice(index, 1)
    tempArr.push(...item)
  })
  tempArr.forEach((item) => {
    originTodoList.value.push(item)
  })
  todoList.value = originTodoList.value
})

watch(isVisibleDone, (value) => {
  todoList.value = originTodoList.value.filter((item) => {
    return value ? item.id : !item.done
  })
})

const handleAddTodo = () => {
  let name = formData.name
  todoService.createTodo({ name }).then((res) => {
    if (res.id) {
      formData.name = ''
      todoService.getTodoList().then((list) => {
        originTodoList.value = list
        ElMessage.success('新增成功')
      })
    }
  })
}

const handleUpdateTodo = (todo: Todo) => {
  todoService
    .update(todo.id, {
      name: todo.name,
      done: Boolean(todo.done),
    })
    .then((res) => {
      console.log(res)
    })
}

const handleUpdateDoneTodo = (todo: Todo) => {
  todoService
    .update(todo.id, {
      done: !todo.done,
    })
    .then((res) => {
      todo.done = res.done
    })
}

const handleDeleteTodo = (todo: Todo) => {
  todoService.delete(todo.id).then(() => {
    ElMessage.info('删除成功')
    todoService.getTodoList().then((list) => {
      originTodoList.value = list
    })
  })
}

const handleVisibleDone = () => {
  isVisibleDone.value = !isVisibleDone.value
}
</script>

<style scoped lang="less">
.input {
  outline: none;
  background: none;
  border: none;
  font-size: 20px;
  color: #fff;
  font-weight: 600;
}

.todo-list {
  width: 400px;
  margin: auto;

  ul,
  li {
    list-style: none;
    padding-left: 0;
  }

  li {
    padding: 20px 20px;
    text-align: left;
    background: rgba(70, 209, 197, 0.5);
    border-radius: 4px;
    position: relative;
    margin-bottom: 20px;

    input:focus {
      border: none;
    }
  }

  .done {
    height: 30px;
    width: 30px;
    background: #42b983;
    border-radius: 50%;
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    right: 60px;
    cursor: pointer;
    text-align: center;
  }

  .delete {
    border-radius: 50%;
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    right: 20px;
    cursor: pointer;
  }

  .todo-done {
    background: gainsboro;
    text-decoration: line-through;

    input {
      text-decoration: line-through !important;
    }
  }
}

.button {
  outline: none;
  border: none;
  cursor: pointer;
  padding: 10px 20px;
  border-radius: 4px;
  background: cornflowerblue;
  color: #fff;
  font-weight: bold;
  font-size: 16px;
  margin-left: 20px;
  opacity: 0.9;
}
</style>
