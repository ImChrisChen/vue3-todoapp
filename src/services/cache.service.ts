interface CacheOptions {
  time?: number
  id?: string
}

interface CacheContent extends CacheOptions {
  value: unknown
}

// 需要使用Key时需要在这里注册
type CacheKey = 'todolist'

class CacheService {
  cacheKey = 'TodoAppCache:'
  defaultExpireTime = 1000 * 60 * 15 // 默认缓存时间15分钟

  private getKey(key: string) {
    return this.cacheKey + key
  }

  setItem(key: CacheKey, value, options: CacheOptions = {}) {
    try {
      const finalKey = this.getKey(key)
      const content: CacheContent = {
        time: (options.time || this.defaultExpireTime) + Date.now(),
        id: options.id || '',
        value: value,
      }
      localStorage.setItem(finalKey, JSON.stringify(content))
      return true
    } catch (e) {
      return false
    }
  }

  getItem(key: CacheKey, options: Omit<CacheOptions, 'time'> = {}) {
    if (!key) {
      new Error('key为必须')
      return null
    }

    let content = {} as CacheContent

    try {
      const finalKey = this.getKey(key)
      const jsonStr: string = localStorage.getItem(finalKey) || '{}'
      content = JSON.parse(jsonStr)
    } catch (e) {
      console.error(e)
    }

    // id mismatch
    if (options.id && content.id && options.id !== content.id) {
      return null
    }

    // availability time
    if (!content.time || content.time < Date.now()) {
      return null
    }

    return content.value
  }

  removeItem(key: string) {
    return localStorage.removeItem(key)
  }
}

export const cacheService = new CacheService()
