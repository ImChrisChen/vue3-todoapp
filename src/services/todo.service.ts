import httpClient, { ResponseData } from '@/http/http'

export declare namespace TodoDataDto {
  interface CreateTodoDto {
    name: string
    done?: boolean
  }

  interface UpdateTodoDto {
    name?: string
    done?: boolean
  }
}

class TodoService {
  async update<T>(id: number, data: TodoDataDto.UpdateTodoDto) {
    return httpClient.put(`/todo/${id}`, data)
  }

  async delete(id: number) {
    return httpClient.delete(`/todo/${id}`)
  }

  async getTodoList<T>(): Promise<T | any[]> {
    return httpClient.get<ResponseData<T>>(`/todo/`).then((res) => {
      if (res.data.result !== 'ok') {
        console.error('')
        // return res.data.data
      }
      return res?.data?.data || []
    })
  }

  async createTodo(data: TodoDataDto.CreateTodoDto) {
    if (!('done' in data)) {
      data.done = false
    }
    if (!data.name) {
      // console.error('name不能为空', data)
      throw new Error(`name不能为空`)
    }

    return httpClient.post<ResponseData<any>>(`/todo`, data).then((res) => {
      return res.data
    })
  }
}

export const todoService = new TodoService()
