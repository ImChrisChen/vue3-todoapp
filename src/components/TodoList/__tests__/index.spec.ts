import TodoList from '@/components/TodoList/index.vue'
import { flushPromises, mount } from '@vue/test-utils'
import { Todo, todoService } from '@/services'

describe('TodoList', () => {
  it('jest.fn', () => {
    const mockFn = jest.fn()
    const result = mockFn(1, 2, 3)

    expect(result).toBeUndefined()
    expect(mockFn).toBeCalled()
    expect(mockFn).toBeCalledTimes(1)
    expect(mockFn).toHaveBeenCalledWith(1, 2, 3)
  })

  it('render todolist', async () => {
    const list: Array<Todo> = [
      {
        id: 1,
        name: '洗澡12313',
        created_at: '2022-07-02T15:55:43.000Z',
        updated_at: '2022-07-14T14:32:48.000Z',
        deadline: null,
        remark: null,
        done: 0,
      },
      {
        id: 2,
        name: '起床',
        created_at: '2022-07-02T17:03:16.000Z',
        updated_at: '2022-07-14T14:32:46.000Z',
        deadline: null,
        remark: null,
        done: 1,
      },
      {
        id: 6,
        name: '11点洗澡',
        created_at: '2022-07-03T14:10:36.000Z',
        updated_at: '2022-07-14T14:32:43.000Z',
        deadline: null,
        remark: '',
        done: 1,
      },
      {
        id: 8,
        name: '提交代码',
        created_at: '2022-07-03T14:29:00.000Z',
        updated_at: '2022-07-14T14:32:39.000Z',
        deadline: null,
        remark: '',
        done: 0,
      },
    ]
    jest.spyOn(todoService, 'getTodoList').mockResolvedValue(list)

    const todoList = mount(TodoList)

    await flushPromises()

    expect(todoList.exists()).toBe(true)
    const l = todoList.findAll('.todo')
    expect(l).toHaveLength(list.length)
    expect(todoList.findAll('.todo.todo-done')).toHaveLength(list.filter((item) => item.done).length)

    // const values = todoList.findAll('.todo .input').map((item) => item.element['value'])
    // expect(values).toEqual(list.map((item) => item.name))
    // expect(values).toContainEqual(values)
    expect(todoList).toMatchSnapshot()
  })
})
