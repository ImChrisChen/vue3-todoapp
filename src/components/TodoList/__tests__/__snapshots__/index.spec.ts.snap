// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList render todolist 1`] = `
"<div class=\\"todo-list\\"><input class=\\"input\\" placeholder=\\"请输入要完成的事项\\" style=\\"background: seashell; color: rgb(44, 62, 80); display: inline-block; border-radius: 4px; padding: 10px;\\" type=\\"text\\"><button class=\\"button\\">添加</button>
  <ul>
    <li class=\\"todo\\"><input class=\\"input\\" type=\\"text\\"><span class=\\"done\\">✔️</span><span class=\\"close\\">❌</span></li>
    <li class=\\"todo\\"><input class=\\"input\\" type=\\"text\\"><span class=\\"done\\">✔️</span><span class=\\"close\\">❌</span></li>
    <li class=\\"todo-done todo\\"><input class=\\"input\\" type=\\"text\\"><span class=\\"done\\">✔️</span><span class=\\"close\\">❌</span></li>
    <li class=\\"todo-done todo\\"><input class=\\"input\\" type=\\"text\\"><span class=\\"done\\">✔️</span><span class=\\"close\\">❌</span></li>
  </ul><button class=\\"button\\">隐藏 已完成</button>
</div>"
`;
